File: apps-microservices/redis-client-frontend/lib/auth/session.ts
import { SignJWT, jwtVerify } from "jose"

export const SESSION_COOKIE = "rcf_session"

function sessionKey(): Uint8Array {
  const secret = process.env.SESSION_SECRET
  if (!secret) throw new Error("[redis-client] Missing SESSION_SECRET")
  return new TextEncoder().encode(secret)
}

export interface SessionClaims {
  email: string
  name?: string
}

export async function createSessionToken(claims: SessionClaims, ttlSeconds: number): Promise<string> {
  return new SignJWT({ name: claims.name })
    .setProtectedHeader({ alg: "HS256" })
    .setSubject(claims.email)
    .setIssuedAt()
    .setExpirationTime(Math.floor(Date.now() / 1000) + ttlSeconds)
    .sign(sessionKey())
}

export async function readSession(token: string | undefined): Promise<SessionClaims | null> {
  if (!token) return null
  try {
    const { payload } = await jwtVerify(token, sessionKey(), { algorithms: ["HS256"] })
    const email = payload.sub
    if (!email) return null
    return { email, name: payload.name as string | undefined }
  } catch {
    return null
  }
}

File: libs/auth/python/hellopro_auth/session.py
"""Signed rcf_session JWT (HS256, SESSION_SECRET). Parity with libs/auth/node session.ts."""
from __future__ import annotations

import os
import time
from dataclasses import dataclass

import jwt

SESSION_COOKIE = "rcf_session"


def _session_secret() -> str:
    secret = os.environ.get("SESSION_SECRET", "").strip()
    if not secret:
        raise RuntimeError("[account-auth] Missing SESSION_SECRET")
    return secret


@dataclass(frozen=True)
class SessionClaims:
    email: str
    name: str | None = None


def create_session_token(claims: SessionClaims, ttl_seconds: int) -> str:
    now = int(time.time())
    payload = {"sub": claims.email, "name": claims.name, "iat": now, "exp": now + ttl_seconds}
    return jwt.encode(payload, _session_secret(), algorithm="HS256")


def read_session(token: str | None) -> SessionClaims | None:
    if not token:
        return None
    secret = _session_secret()  # raise loudly on misconfig, before the try
    try:
        payload = jwt.decode(token, secret, algorithms=["HS256"])
    except jwt.PyJWTError:
        return None
    email = payload.get("sub")
    if not email:
        return None
    return SessionClaims(email=email, name=payload.get("name"))
